docs: CA-6191 doc polish only; no code change (#119)
ReachFive/reachfive-ios: docs/modules/ROOT/pages/guides/custom-provider.adoc

= Implement a custom provider

Our built-in xref:providerCreator.adoc[provider creators] cover the most common use cases for provider creation:

* `GoogleProvider`
* `FacebookProvider`
* `AppleProvider`
* `WeChat`
* `WebProvider`

This custom provider guide instructs you how to implement a provider that none of the built-ins model such as a social/SSO provider with its own native SDK, or a login flow that none of the built-ins model.

If your provider has no native SDK component and only needs to choose *how* its `ASWebAuthenticationSession` returns to the app (custom scheme, universal link, out-of-band), you don't need a custom provider, instead configure a xref:providerCreator.adoc[`WebProvider`].

== The two protocols

A provider is made of two pieces, both found in `Sources/Core/Classes/Provider.swift`:

* `ProviderCreator`: a lightweight factory you list in `ReachFive(providersCreators:)`. 
It carries a `name` and an optional `variant` and builds the actual `Provider` once {company} has fetched that provider's configuration from the backend.

> Its `create` method also receives the shared `ReachFive` instance.

Reviewer: Also {company}?

```suggestion
Its `create` method also receives the shared `{company}` instance.
```

See <<Example: wrapping a native SDK>> for how to hold that reference safely in the `Provider` you return.
* `Provider`: the object doing the actual work: `login`, `logout`, and the app-lifecycle hooks it needs to intercept its own callback.
+
[source,swift]
----
include::example$customProviderProtocols.swift[]
----

IMPORTANT: `name` must match the provider's name as configured on your {cc}.
This is how {company} matches your `ProviderCreator` against the provider configurations it fetches from the backend.

> A configured provider with no matching creator (and that isn't Apple) still logs in, falling back to a generic web-based `DefaultProvider`.

Reviewer: (this was my phrasing, but) Maybe remove the "and that isn't Apple" aside and say that it falls back either to web-based `DefaultProvider` or the native Sign in with Apple flow.
I think this would feel safer

```suggestion
A configured provider with no matching creator still logs in, falling back to either native Sign in with Apple or a generic web-based provider.
```

== Example: wrapping a native SDK

This is the shape used by our own (link:https://github.com/ReachFive/reachfive-ios/blob/master/Sources/Core/Classes/AppleProvider.swift[`AppleProvider.swift`^]): 

* The creator is a thin factory.
* The actual `Provider` calls into the third-party native SDK, then exchanges its result for a {company} `AuthToken` through `reachFive.authWithCode`, which is the same helper used internally by native login.

Your `create` method receives the shared `ReachFive` instance, which the SDK also stores in `ReachFive.providers`.
When you build your `Provider`, decide how to access it from `login`:

[CAUTION]
====
When `create(reachFive:)` runs, the SDK has already decided to keep your `Provider` in `ReachFive.providers`.

> If your provider also keeps a *strong* reference back to that same `ReachFive` instance, neither object can ever be released—a *retain cycle* (`ReachFive` ↔ `Provider`) that pins the whole SDK object graph in memory.

Reviewer: ```suggestion
If your provider also keeps a *strong* reference back to that same `ReachFive` instance, neither object can ever be released — a *retain cycle* (`ReachFive` ↔ `Provider`) that pins the whole SDK object graph in memory.
```

Use `weak var reachFive` in your provider, or copy only the properties you need at init time (e.g. `sdkConfig`, `reachFiveApi`, `scope`).

> For background on why this happens and how `weak` breaks the cycle, see link:https://developer.apple.com/documentation/swift/automatic-reference-counting#Strong-Reference-Cycles-Between-Class-Instances[Apple: strong reference cycles between class instances^] and link:https://developer.apple.com/documentation/swift/weak[`weak` references^].

Reviewer: Fix broken links
```suggestion
For background on why this happens and how `weak` breaks the cycle, see link:https://docs.swift.org/swift-book/documentation/the-swift-programming-language/automaticreferencecounting/#Strong-Reference-Cycles-Between-Class-Instances[Apple: strong reference cycles between class instances^] and link:https://docs.swift.org/swift-book/documentation/the-swift-programming-language/automaticreferencecounting/#Weak-References[`weak` references^].
```
